### PR TITLE
Filter Timeout Change Refactor

### PR DESCRIPTION
**Pull request by nicholasguoalgorand**
## Summary

https://github.com/algorand/go-algorand/issues/1505
Change new filtertimeout variables to AgreementFilterTimeout and AgreementFilterTimeoutPeriod0 with type time.Duration

## Test Plan

Reran unit tests

### CODE REVIEW

File: agreement/types.go [hunk 1]
```diff
@@ -31,13 +31,10 @@ var recoveryExtraTimeout = config.Protocol.SmallLambda
...
 func FilterTimeout(p period, v protocol.ConsensusVersion) time.Duration {
-	var smallLambdas uint64
 	if p == 0 {
-		smallLambdas = config.Consensus[v].FilterTimeoutPeriod0SmallLambdas
-	} else {
-		smallLambdas = config.Consensus[v].FilterTimeoutSmallLambdas
+		return config.Consensus[v].AgreementFilterTimeoutPeriod0
 	}
-	return config.Protocol.SmallLambda * time.Duration(smallLambdas)
+	return config.Consensus[v].AgreementFilterTimeout
```
Comment (tsachiherman) [inline agreement/types.go:37]: Add a comment here stating that the timeout is expected to be 2 * SmallLambda, and it was moved into the consensus protocol parameter.

File: config/consensus.go [hunk 2]
```diff
@@ -137,8 +137,8 @@ type ConsensusParams struct {
 	DownCommitteeSize      uint64
 	DownCommitteeThreshold uint64
 
-	FilterTimeoutSmallLambdas             uint64
-	FilterTimeoutPeriod0SmallLambdas      uint64
+	AgreementFilterTimeout        time.Duration
+	AgreementFilterTimeoutPeriod0 time.Duration
```
Comment (tsachiherman) [inline config/consensus.go:141]: Add explanations around these variables - what do they mean, how they should be configured, etc.

File: config/consensus.go [hunk 3]
```diff
@@ -792,7 +792,7 @@ func initConsensusProtocols() {
 	vFuture.ApprovedUpgrades = map[protocol.ConsensusVersion]uint64{}
 
 	// FilterTimeout 2s instead of 4s for period 0
-	vFuture.FilterTimeoutPeriod0SmallLambdas = 1
+	vFuture.AgreementFilterTimeoutPeriod0 = 2 * time.Second
```
Comment (tsachiherman) [inline config/consensus.go:795]: for now, set it to `4 * time.Seconds` and update the comment that we should revisit this value.

**Review by tsachiherman — COMMENTED** [review]
Few notes below; In addition, take a look on `node_test.go` : 
the following line should be updated:

var expectedAgreementTime = 2*config.Protocol.BigLambda + 3*config.Protocol.SmallLambda + 2*time.Second

The  3*config.Protocol.SmallLambda should be converted to `config.Protocol.SmallLambda + cparams.agreementfiltertimeout`

**Comment by nicholasguoalgorand** [comment]
> Few notes below; In addition, take a look on
> 
> 1. `node_test.go` :
>    the following line should be updated:
> 
> var expectedAgreementTime = 2_config.Protocol.BigLambda + 3_config.Protocol.SmallLambda + 2*time.Second
> 
> The 3*config.Protocol.SmallLambda should be converted to `config.Protocol.SmallLambda + cparams.agreementfiltertimeout`
> 
> 1. `basic_test.go`:
>    this line `time.Sleep(2 * config.Protocol.SmallLambda)` need to be updated to be `cparams.agreementfiltertimeout`
> 
> once you're done with this change, you'll have some more work refactoring the `ALGOSMALLLAMBDAMSEC` into the `agreementfiltertimeout` that you just created.

What consensus version should `cparams` be, current or future?

Next for the `ALGOSMALLLAMBDAMSEC` refactor, is it okay to just create env variables for `ALGOFILTERTIMEOUT`?

**Comment by tsachiherman** [comment]
> > Few notes below; In addition, take a look on
> > 
> > 1. `node_test.go` :
> >    the following line should be updated:
> > 
> > var expectedAgreementTime = 2_config.Protocol.BigLambda + 3_config.Protocol.SmallLambda + 2*time.Second
> > The 3*config.Protocol.SmallLambda should be converted to `config.Protocol.SmallLambda + cparams.agreementfiltertimeout`
> > 
> > 1. `basic_test.go`:
> >    this line `time.Sleep(2 * config.Protocol.SmallLambda)` need to be updated to be `cparams.agreementfiltertimeout`
> > 
> > once you're done with this change, you'll have some more work refactoring the `ALGOSMALLLAMBDAMSEC` into the `agreementfiltertimeout` that you just created.
> 
> What consensus version should `cparams` be, current or future?
> 
> Next for the `ALGOSMALLLAMBDAMSEC` refactor, is it okay to just create env variables for `ALGOFILTERTIMEOUT`?

Definitely use params with current version. This test is not intended to test the variations of the filtertimeout, so using the current protocol version is the correct thing to do.

As for the ALGOSMALLLAMBDAMSEC refactor, no - it should be a configurable consensus where you programmatically set the FilterTimeout. The fact that we had that ALGOSMALLLAMBDAMSEC variable was a temporary placeholder. ( that small refactor should be done outside the scope of this PR )

File: test/e2e-go/features/transactions/accountv2_test.go [hunk 4]
```diff
@@ -39,6 +40,8 @@ func TestAccountInformationV2(t *testing.T) {
 	proto, ok := config.Consensus[protocol.ConsensusFuture]
 	a.True(ok)
 	os.Setenv("ALGOSMALLLAMBDAMSEC", "200")
+	proto.AgreementFilterTimeoutPeriod0 = 400 * time.Millisecond
+	proto.AgreementFilterTimeout = 400 * time.Millisecond
```
Comment (tsachiherman) [inline test/e2e-go/features/transactions/accountv2_test.go:44]: if you've set the proto.AgreementFilterTimeoutPeriod0/proto.AgreementFilterTimeout, than you should be able to remove the `os.Setenv("ALGOSMALLLAMBDAMSEC", "200")` and ```defer func() {
 		os.Unsetenv("ALGOSMALLLAMBDAMSEC")
	}()```, right ?

File: test/e2e-go/upgrades/send_receive_upgrade_test.go [hunk 5]
```diff
@@ -117,6 +117,10 @@ func generateFastUpgradeConsensus() (fastUpgradeProtocols config.ConsensusProtoc
 		}
 
 		fastUpgradeProtocols[consensusTestFastUpgrade(proto)] = fastParams
+
+		// support the ALGOSMALLLAMBDAMSEC = 500 env variable
+		fastParams.AgreementFilterTimeout = time.Second
```
Comment (tsachiherman) [inline test/e2e-go/upgrades/send_receive_upgrade_test.go:122]: as before - if you've set this, you should be able to get rid of the `os.Setenv("ALGOSMALLLAMBDAMSEC", "500")`.
make sure that the various tests still takes have the same execution time.

**Review by tsachiherman — APPROVED** [review]
@nicholasguoalgorand , the changes here are good, but its not complete yet.
I'm going to merge this in just to unblock the performance testing, but the rest of the `ALGOSMALLLAMBDAMSEC` still need to be updated.

**Comment by nicholasguoalgorand** [comment]
I wasn't sure if the env variables could be removed since small lambda is still used in other variables ie deadlineTimeout, etc

**Comment by tsachiherman** [comment]
True, but these tests are only adjusting the small lambda to run faster,
and it would suffice to update the filtertimeout in these cases.
( we don't have any tests that arbitrarily check various small lambda
values )

**Comment by nicholasguoalgorand** [comment]
Cool, then I can make another pr that removes those? We might be able to completely deprecate the env variable?

**Comment by tsachiherman** [comment]
Yes, that was my intent.

**Pull request merged**